implementation/src/main/java/io/smallrye/config/NameIterator.java

```diff
@@ -7,7 +7,7 @@
 /**
  * An iterator for property name strings.
  */
-public final class NameIterator {
+public final class NameIterator implements AutoCloseable {
```

Comment:
I don't really like the stylistic change here of making the NI be closeable; does the lifespan of the NI really leak out enough that the string needs to be cleared?

Reply:
It's not to clear the String, but to avoid re allocating the buffer underlying the StringBuilder used to extract the segment. Sadly I have performed a similar optimization on Jandex, for a similar reason..
Escape analysis kick in on C2 and still, even if it does it, it won't make the builder to be scalar replaced (because its underlying buffer can be enlarged, so...cannot be constant folded :/)
The reason why I made it closeable is to leverage the try with resource and reset the underlying state, to not mess up with subsequent reuse

Reply:
Anyway, the change I did to make getNextSegment able to use an external builder, could be enough to make it immutable again...but I have the strong feeling that other core path cannot leverage the same new API and would benefit from reusingnthe underlying builder...I am open to any other idea to improve it/make it right.

Pooling usually suck, API wise, due to introduced mutability

Reply:
@dmlloyd are you ok with this?